Add task-based history export tracking

davelopez opened the pull request:
Requires #15090 (To pass the remaining integration test failure)
Fixes #14511
Related to #14771

## TLDR

This PR enables the tracking of every history export (and potentially any other exportable object) in the database by adding a new table containing structured metadata about the export context.
It also creates a new `HistoryExport` component in the UI that is based on the asynchronous task (Celery) system instead of the regular job system for handling exports and short-term downloads (#14511).

## Backend Changes

- Adds a new `store_export_association` table to track any object export associated with a task. This table has an `export_metadata` column containing information about the export context (request and result) in a structured JSON (modeled with Pydantic objects).

- Adds a new `StoreExportTracker` class to associate export metadata with any exportable object.

- Refactors the `HistoryExportView` to `HistoryExportManager` for managing the exports associated with histories.

- Adapts the `model_stores` to generate and associate the export metadata to history downloads and remote exports.

- Refactors the Tasks API to use an abstract `AsyncTaskManager` and the corresponding `CeleryAsyncTaskManager` implementation to query tasks status.

- Modifies the `/api/histories/{id}/exports` endpoint to accept a new type `application/vnd.galaxy.task.export+json` that will return the new task-based export associations. It will continue returning the previous "JEHAs" using the default request accept type so it's backward compatible.

- Some other small refactorings for the schemas, removing unused code, etc.

## Frontend Changes

Added several new components to display information on the last export and previous ones along with actions to download them again or re-import them from remote sources, or even choose the preferred archive format. They provide similar functionality as the previous components (that still are used when Celery is not enabled) and include a bit more related to the export tracking.

### Choose Export Options

Now you can decide which export format and dataset files want to include in the archive.

![image](https://user-images.githubusercontent.com/46503462/205378170-7e625ea5-b3c5-4d9f-b377-9b3f88ed7b12.png)


### Direct Download (Temporal)

When you use the direct download option, it will be reused (the package won't be regenerated) whenever you try to download it again until:

- The download link expires
- The history changes
- The export parameter change

  ![image](https://user-images.githubusercontent.com/46503462/203846184-7c64c252-6a5d-4792-92ed-d377e76d6fab.png)

### Export to Remote File Source (Long term)

When you export your history to a remote file source a reference will be stored in the record and you can reimport it later as long as the remote file source maintains the exported package.
![image](https://user-images.githubusercontent.com/46503462/203846392-6b9e8cad-ab1a-4709-9fbd-880a792a6cf5.png)

### List of previous exports

You can download or reimport copies (snapshots) of previous versions of the history as long as they have not expired or the remote source maintains them.
![image](https://user-images.githubusercontent.com/46503462/203847316-3fc0b0c4-2cb6-44b0-a242-47431fc25e20.png)

## TODO

- [x] Display options to select/change the export format
- [x] Add a bunch of tests
- [ ] Make the UI reactive to history changes (requires history store (Pinia) for composition API)

## How to test the changes?

(Select all options that apply)

- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License

- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Add `StoreExportAssociation` table

Used to track exported objects (currently histories and invocations) including additional metadata.

davelopez committed:
Add abstract AsyncTasksManager

And celery implementation in CeleryAsyncTasksManager

davelopez committed:
Allow registering concrete classes for interfaces in DI

davelopez committed:
Use AsyncTasksManager in Tasks API

davelopez committed:
Add StoreExportTracker

Manages export-related metadata association of different types of exportable objects to remote file sources or repositories.

davelopez commented on lib/galaxy/schema/schema.py:
```diff
@@ -1410,8 +1414,40 @@ class JobExportHistoryArchiveModel(Model):
+
 class JobExportHistoryArchiveCollection(Model):
-    __root__: List[JobExportHistoryArchiveModel]
+    __root__: List[AnyExportHistoryArchiveModel]
```
I'm trying to decide if returning both types of "views" (Job or Task) of a history export will be too much of a breaking change for the API...

Will it be more desirable to create a dedicated route for the task-based exports (like `api/histories/{id}/exports/tasks`) and leave the current `api/histories/{id}/exports` just for the job-based one so the API doesn't change that much?
OTOH I feel like it will be a bit harder to maintain in the client later if you need to request both endpoints to know how the history was exported... :thinking:

davelopez replied:
I guess the question boils down to... what is the best approach to associate export metadata to an exported "object" (not necessarily only histories) so we can track when this object was exported and where (for recovery)?

I was hoping to use `api/histories/{id}/exports` for the particular case of histories because it makes a lot of sense semantically. But maybe we should use a more general API independent of the object type?

davelopez replied:
Following up with the discussion yesterday in the backend meeting about this...

1 - Is it a better idea to create a new `GET /api/histories/{id}/store_exports` endpoint instead of trying to reuse `GET /api/histories/{id}/exports`? this **store_exports** endpoint will have an equivalent for all objects that support the **write_store** and **from_store** endpoints (like invocations)

2 - Assuming we want to use case (1). How do we want to persist the export metadata for all these objects (histories, invocations, etc)? Creating a new "store_export_association" table that tracks all the exports (excluding simple downloads) for all types of exported objects? 

3 - Do we want to deprecate the job-based export for histories and use only the task-based one at some point? Meaning we should update the current history export UI to align it more with #14606?

mvdbeek replied:
> Creating a new "store_export_association" table that tracks all the exports (excluding simple downloads) for all types of exported objects?

I think that's a solid yes.

> this **store_exports** endpoint will have an equivalent for all objects that support the **write_store** and **from_store** endpoints (like invocations)

You mean we'll also have `/api/invocations/{id}/store_exports` ? This sounds good to me, my one concern is that `store` feels like an implementation detail consumers wouldn't need to know about. Is there any chance we can gloss over the differences and keep a backwards compatible API at `/api/histories/{id}/exports` ?

> Do we want to deprecate the job-based export for histories and use only the task-based one at some point? Meaning we should update the current history export UI to align it more with #14606?

Yes IMO. That would close https://github.com/galaxyproject/galaxy/issues/14511

davelopez replied:
Thank you @mvdbeek, that helps a lot!

> You mean we'll also have /api/invocations/{id}/store_exports

Yes, that is what I meant :)

> Is there any chance we can gloss over the differences and keep a backwards compatible API at /api/histories/{id}/exports

The only solution that pops into my mind is using a custom content-type for returning just the task-based exports and fallback to the job-based one with the default content-type. Will that be OK in this case?

> Yes IMO. That would close https://github.com/galaxyproject/galaxy/issues/14511

Oh! great thanks! I lost track of this one :)

davelopez replied:
I'm happy now using a different accept content type to make the `/api/histories/{id}/exports` compatible with both export modes, it feels cleaner and works great with the interactive API docs :)

davelopez committed:
Remove unused dependency in HistoryController

davelopez committed:
Use StoreExportTracker with histories

Associate basic export metadata when exporting to a store

davelopez committed:
Use custom accept content type in exports API

To make the `/api/histories/{id}/exports` endpoint compatible with both exports methods.
By default the jeha exports are returned for retro-compatibility.
The new task-based exports with metadata are returned when using `application/vnd.galaxy.task.export+json` as accept content type.

davelopez committed:
Move ExportObjectType enum to schema

davelopez committed:
Update history exports API docs

davelopez committed:
Order exports by date

davelopez committed:
Refactor some schema model names

davelopez committed:
Add FileSources composable

davelopez committed:
Add pagination support to exports endpoint

davelopez committed:
Deprecate jeha-based export API endpoints

davelopez committed:
Allow configuring polling time in useTaskMonitor

davelopez committed:
Add basic task based history export UI

davelopez committed:
Add ShortTermStorage composable

davelopez committed:
Add basic history direct download using STS

davelopez committed:
Refactor schemas for short term export models

davelopez committed:
Refactor rename variables for consistency

davelopez committed:
Track direct downloads to reuse the STS export

davelopez committed:
Use correct model_store_format when importing from uri

davelopez committed:
Implement history reimport from file source

davelopez committed:
Fix imports order

davelopez committed:
Add re-import confirmation

davelopez committed:
Add STS duration to ShortTermStorageTarget

This simplifies tracking the expiration date of a temporarily stored object.

davelopez committed:
Add expiration date to short term download links

davelopez committed:
Refactor export record details date management

davelopez committed:
Add basic table with previous export records

davelopez committed:
Disable celery in existing history export selenium test

So the previous job-based export system is correctly tested

davelopez committed:
Refactor loading of records to make single request

davelopez committed:
Reuse up to date download record if available

davelopez committed:
Add more information and refactor messages

davelopez committed:
Optionally clear ExportForm input after export

davelopez committed:
Add more test coverage to ExportForm

davelopez committed:
Add ExportOptions component

davelopez committed:
Display export format in records

davelopez committed:
Simplify export result handling

By caching the result in the metadata and not relying on task status queries

davelopez committed:
Various fixes around export parameters

- Check export params are the same before re-downloading otherwise generate a new download
- Fix display status when download has expired
-  Minor refactoring

davelopez committed:
Add some jest tests

davelopez committed:
Add API test for history export tracking

davelopez committed:
Fix mypy ignore type after version update

davelopez committed:
Increase `test_export_tracking` coverage

davelopez committed:
Add integration selenium test for export tracking

davelopez committed:
Fix export record `ready` state + refactor test

davelopez committed:
Convert service functions to TS

davelopez committed:
Display detailed error message if the export fails

davelopez committed:
Rename path parameter id to history_id

This is more explicit and consistent, also plays nicely with the client generated code.
- Remove redundant docstring in service.
- Update client generated schema.

davelopez committed:
Migrate ExportRecordModel to typescript

davelopez committed:
Migrate history export services to typescript

And use the generated OpenAPI types.

davelopez committed:
Cast the response type in getExportRecords

Unfortunately the response is always inferred from the default content type 'application/json' so we need to change it to 'unknown' and the cast it to the expected type manually.

davelopez committed:
Migrate jest tests to typescript

jmchilton commented on lib/galaxy_test/selenium/test_history_export.py:
```diff
@@ -7,6 +7,10 @@
 
+    @classmethod
+    def handle_galaxy_config_kwds(cls, config):
+        config["enable_celery_tasks"] = False
```
Non-integration tests shouldn't override this. Previously I don't think they even had this method. Is this definitely being used?

davelopez replied:
> Is this definitely being used?

Yes, it does work, see the call stack in the screenshot to find where is being called:
![image](https://user-images.githubusercontent.com/46503462/206753876-fa1798da-e158-4a58-b7ed-9760bd8238e0.png)

But I see your point... this kind of turned into an integration test with this change... should I just move it to test/integration_selenium then?

davelopez replied:
Since Celery is enabled by default in the test instance I need it to be disabled to keep testing the previous UI, so I decided to move it and convert it into an integration test for consistency.

Thank you John! let me know if I should approach it differently.

jmchilton commented on lib/galaxy/webapps/galaxy/api/histories.py:
```diff
@@ -379,19 +412,23 @@ def archive_export(
         code (instead of 202) and this route can be used to download the archive.
+
+        **Deprecation notice**: Please use `/api/histories/{id}/prepare_store_download` or
+        `/api/histories/{id}/write_store` instead.
```
❤️

jmchilton commented on lib/galaxy/webapps/galaxy/api/tasks.py:
```diff
@@ -14,11 +17,12 @@
 
 @router.cbv
 class FastAPITasks:
+    manager: AsyncTasksManager = depends(AsyncTasksManager)  # type: ignore[type-abstract]
```
Over and over mypy just sucks on this point - I want to ``--disable-error-code=type-abstract`` this whole project. xref https://github.com/python/mypy/issues/4717#issuecomment-1264388209. I don't think I've ever seen a bug found via that issue - it just makes good code and code patterns look ugly.

davelopez replied:
Yeah, it's a bit disappointing that mypy doesn't handle this nicely...

jmchilton commented on lib/galaxy/app.py:
```diff
@@ -538,6 +542,10 @@ def __init__(self, configure_logging=True, use_converters=True, use_display_appl
         self.role_manager = self._register_singleton(RoleManager)
         self.job_manager = self._register_singleton(JobManager)
 
+        self.task_manager = self._register_abstract_singleton(
```
Wonderful! I love this!

davelopez committed:
Add back ExportParamsModel class

davelopez committed:
Move test_history_export selenium test

This selenium test is for testing the previous history export UI that is based on jobs and not tasks.
Since Celery is enabled by default in the test instance it is more consistent to convert it into an integration selenium test to modify the config and disable Celery.

mvdbeek reviewed (approved):
Absolutely amazing work, thanks so much @davelopez!
